File: core/src/session_state.cpp
#include "hearport/session_state.h"

namespace hearport {

bool SessionState::ReceiveConnect(AuthMode mode,
                                  std::span<const std::byte> peer_id) noexcept {
  if (phase_ != SessionPhase::awaiting_connect ||
      !ValidateConnectRequest(mode, peer_id)) {
    phase_ = SessionPhase::closed;
    return false;
  }
  auth_mode_ = mode;
  phase_ = SessionPhase::authenticating;
  return true;
}

bool SessionState::MarkAuthenticated() noexcept {
  if (phase_ != SessionPhase::authenticating) {
    return false;
  }
  phase_ = SessionPhase::ready;
  return true;
}

bool SessionState::BeginStream(std::uint32_t stream_id) noexcept {
  if (stream_id == 0 ||
      (phase_ != SessionPhase::ready && phase_ != SessionPhase::active)) {
    return false;
  }
  pending_stream_id_ = stream_id;
  active_stream_id_.reset();
  phase_ = SessionPhase::pending_stream;
  return true;
}

AudioDisposition SessionState::AckWritten(std::uint32_t stream_id) noexcept {
  if (phase_ != SessionPhase::pending_stream ||
      pending_stream_id_.value_or(0) != stream_id || stream_id == 0) {
    return AudioDisposition::protocol_error;
  }
  active_stream_id_ = stream_id;
  pending_stream_id_.reset();
  phase_ = SessionPhase::active;
  return AudioDisposition::accepted;
}

AudioDisposition SessionState::AcceptAudio(
    const wire::AudioDatagram& packet) const noexcept {
  if (phase_ == SessionPhase::pending_stream) {
    return AudioDisposition::pending_audio_discarded;
  }
  if (phase_ == SessionPhase::active &&
      active_stream_id_.value_or(0) == packet.stream_id) {
    return AudioDisposition::accepted;
  }
  return AudioDisposition::old_stream_discarded;
}

bool SessionState::ResetStream(std::uint32_t stream_id) noexcept {
  return BeginStream(stream_id);
}

}  // namespace hearport

File: tests/reference/session_state.py
class ProtocolError(ValueError):
    pass


class SessionState:
    def __init__(self):
        self.phase = "awaiting_connect"
        self.auth_mode = None
        self.pending_stream_id = None
        self.active_stream_id = None

    def receive(self, message: dict) -> str:
        if self.phase != "awaiting_connect":
            raise ProtocolError("ConnectRequest is only legal as the first message")
        if message.get("type") != "connect_request":
            raise ProtocolError("first message must be ConnectRequest")
        auth_mode = message.get("auth_mode")
        peer_id = message.get("peer_id", b"")
        if auth_mode not in {"REMEMBERED", "PAIR", "ONE_TIME"}:
            raise ProtocolError("AUTH_MODE_UNSPECIFIED is invalid")
        if auth_mode == "REMEMBERED" and len(peer_id) != 16:
            raise ProtocolError("remembered peer_id must be exactly 16 bytes")
        if auth_mode != "REMEMBERED" and peer_id:
            raise ProtocolError("pair/one-time peer_id must be empty")
        self.auth_mode = auth_mode
        self.phase = "authenticating"
        return "accepted"

    def mark_authenticated(self) -> None:
        if self.phase != "authenticating":
            raise ProtocolError("authentication is not pending")
        self.phase = "ready"

    def begin_stream(self, stream_id: int) -> None:
        if self.phase not in {"ready", "active"}:
            raise ProtocolError("stream cannot start before authentication")
        if stream_id == 0:
            raise ProtocolError("stream_id must be non-zero")
        self.pending_stream_id = stream_id
        self.active_stream_id = None
        self.phase = "pending_stream"

    def ack_written(self, stream_id: int) -> str:
        if self.phase != "pending_stream" or stream_id != self.pending_stream_id:
            return "protocol_error"
        self.active_stream_id = stream_id
        self.pending_stream_id = None
        self.phase = "active"
        return "active"

    def accept_audio(self, packet) -> str:
        if self.phase == "pending_stream":
            return "pending_audio_discarded"
        if self.phase == "active" and packet.stream_id == self.active_stream_id:
            return "accepted"
        return "old_stream_discarded"

    def reset_stream(self, stream_id: int) -> None:
        self.begin_stream(stream_id)
